src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.cc

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2018- The Pixie Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <libbson-1.0/bson.h>
+
+#include <rapidjson/document.h>
+#include <string>
+
+#include "src/common/base/base.h"
+#include "src/common/base/utils.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/decode.h"
+#include "src/stirling/source_connectors/socket_tracer/protocols/mongodb/types.h"
+#include "src/stirling/utils/binary_decoder.h"
+
+namespace px {
+namespace stirling {
+namespace protocols {
+namespace mongodb {
+
+ParseState ProcessOpMsg(BinaryDecoder* decoder, Frame* frame) {
+  PX_ASSIGN_OR(uint32_t flag_bits, decoder->ExtractLEInt<uint32_t>(), return ParseState::kInvalid);
+
+  // Find relevant flag bit information and ensure remaining bits are not set.
+  // Bits 0-15 are required and bits 16-31 are optional.
+  if (flag_bits & checksum_bitmask) {
+    frame->checksum_present = true;
+  }
+  if (flag_bits & more_to_come_bitmask) {
+    frame->more_to_come = true;
+  }
+  if (flag_bits & exhaust_allowed_bitmask) {
+    frame->exhaust_allowed = true;
+  }
+  if ((flag_bits >> 2) & required_unset_bitmask) {
+    return ParseState::kInvalid;
+  }
+
+  // Determine the number of checksum bytes in the buffer.
+  uint8_t checksum_bytes = (frame->checksum_present) ? 4 : 0;
+
+  // Get the section(s) data from the buffer.
+  while (decoder->BufSize() > checksum_bytes) {
```

Comment:
I believe this will fail if the buffer has more than a full frame of data. We should continue parse at most `frame->length - checksum_bytes` bytes.

In order to test this out, I added additional data to the payload within `MongoDBParserTest.ParseFrameValidRequest` and the test fails rather than parsing the first complete frame and leaving the buffer with the excess data.

```
$ git diff
diff --git a/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.cc b/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.cc
index e9ffa072da..14f1ddd733 100644
--- a/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.cc
+++ b/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.cc
@@ -249,6 +249,11 @@ constexpr uint8_t mongoDBValidRequest[] = {
     0x39, 0x8b, 0x75, 0x41, 0x97, 0xc4, 0x97, 0x06, 0xd1, 0x00, 0x02,
     0x24, 0x64, 0x62, 0x00, 0x06, 0x00, 0x00, 0x00, 0x6d, 0x79, 0x64,
     0x62, 0x31, 0x00, 0x00,
+    // Bogus data
+    0x64, 0x65, 0x72, 0x65, 0x64, 0x00, 0x01, 0x03, 0x6c, 0x73, 0x69,
+    0x64, 0x00, 0x1e, 0x00, 0x00, 0x00, 0x05, 0x69, 0x64, 0x00, 0x10,
+    0x00, 0x00, 0x00, 0x04, 0xe7, 0xd7, 0x16, 0xb3, 0x75, 0xb7, 0x4c,
+    0x39, 0x8b, 0x75, 0x41, 0x97, 0xc4, 0x97, 0x06, 0xd1, 0x00, 0x02,
 };

 constexpr uint8_t mongoDBValidResponse[] = {
$ bazel test --test_output=all --spawn_strategy=standalone src/stirling/source_connectors/socket_tracer/protocols/mongodb:all --test_arg='--gtest_filter=MongoDBParserTest.ParseFrameValidRequest'
INFO: Invocation ID: 63d705ed-b34c-4853-b17f-372237ba2e7c
INFO: Streaming build results to: https://bb.corp.pixielabs.ai/invocation/63d705ed-b34c-4853-b17f-372237ba2e7c
INFO: Analyzed 3 targets (0 packages loaded, 552 targets configured).
INFO: Found 2 targets and 1 test target...
FAIL: //src/stirling/source_connectors/socket_tracer/protocols/mongodb:parse_test (see /home/ddelnano/.cache/bazel/_bazel_ddelnano/fa8604ca46879b9b1ffd44d337c531f3/execroot/px/bazel-out/k8-fastbuild/testlogs/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test/test.log)
INFO: From Testing //src/stirling/source_connectors/socket_tracer/protocols/mongodb:parse_test:
==================== Test output for //src/stirling/source_connectors/socket_tracer/protocols/mongodb:parse_test:
I20230926 22:28:49.784341 1475921 env.cc:47] Started: /home/ddelnano/.cache/bazel/_bazel_ddelnano/fa8604ca46879b9b1ffd44d337c531f3/execroot/px/bazel-out/k8-fastbuild/bin/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.runfiles/px/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test
Note: Google Test filter = MongoDBParserTest.ParseFrameValidRequest
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from MongoDBParserTest
[ RUN      ] MongoDBParserTest.ParseFrameValidRequest
src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test.cc:439: Failure
Expected equality of these values:
  state
    Which is: 4-byte object <01-00 00-00>
  ParseState::kSuccess
    Which is: 4-byte object <05-00 00-00>
[  FAILED  ] MongoDBParserTest.ParseFrameValidRequest (0 ms)
[----------] 1 test from MongoDBParserTest (0 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (0 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] MongoDBParserTest.ParseFrameValidRequest

 1 FAILED TEST
I20230926 22:28:49.784876 1475921 env.cc:51] Shutting down
================================================================================
INFO: Elapsed time: 0.376s, Critical Path: 0.03s
INFO: 2 processes: 1 internal, 1 local.
//src/stirling/source_connectors/socket_tracer/protocols/mongodb:parse_test FAILED in 0.0s
  /home/ddelnano/.cache/bazel/_bazel_ddelnano/fa8604ca46879b9b1ffd44d337c531f3/execroot/px/bazel-out/k8-fastbuild/testlogs/src/stirling/source_connectors/socket_tracer/protocols/mongodb/parse_test/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Streaming build results to: https://bb.corp.pixielabs.ai/invocation/63d705ed-b34c-4853-b17f-372237ba2e7c
INFO: Build completed, 1 test FAILED, 2 total actions
```

With the bogus test data appended, the parser should return a valid frame followed by `ParseState::kInvalid`